### ISSUE TITLE
[BUG] Converting FLOAT to UINT16 causes segfault

### ISSUE THREAD

**Issue by apetrynet** [OpenImageIO/oiio#2880]
**Describe the bug**
In python, when attempting to write a `FLOAT` buffer to disk that has set write format to `UINT16` I get a segfault.
 
**To Reproduce**
``` python
import OpenImageIO as oiio
roi = oiio.ROI(0, 1920, 0, 1080, 0, 1, 0, 3)
buf = oiio.ImageBufAlgo.fill((0, 0, 0), roi=roi)
buf.set_write_format(oiio.UINT16)
buf.write('uint16.dpx')
```
I tested the equivalent code in oiio1.8.x it works.

**Expected behavior**
I expect OIIO to write a dpx file in UINT16

**Evidence**
``` bash
Python 3.8.5 (default, Jan 27 2021, 15:41:15) 
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import OpenImageIO as oiio
>>> roi = oiio.ROI(0, 1920, 0, 1080, 0, 1, 0, 3)
>>> buf = oiio.ImageBufAlgo.fill((0, 0, 0), roi=roi)
>>> buf.set_write_format(oiio.UINT16)
>>> buf.write('uint16.dpx')
Segmentation fault (core dumped)
```
``` bash
Python 2.7.5 (default, Nov 16 2020, 22:23:17) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-44)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import OpenImageIO as oiio
>>> roi = oiio.ROI(0, 1920, 0, 1080, 0, 1, 0, 4)
>>> buf = oiio.ImageBufAlgo.fill((0, 0, 0, 1), roi=roi)
>>> buf.set_write_format(oiio.UINT16)
>>> buf.write('uint16.dpx')
Segmentation fault
```

**Platform information:**
 - OIIO branch/version: 2.2.10.0 (Ubuntu python3.8) and 2.2.10.1 (CentOS7 python2.7)
 - OS: Ubuntu20.04/CentOS7
 - C++ compiler: 9.3.0 (Ubuntu) and 4.8.5 (CentOS7)

**Comment by apetrynet** [comment]
I found a workaround that works in the meantime.

``` python
import OpenImageIO as oiio
roi = oiio.ROI(0, 1920, 0, 1080, 0, 1, 0, 3)
buf = oiio.ImageBufAlgo.fill((1, 0, 0), roi=roi)
obuf = buf.copy(oiio.UINT16)
obuf.write('uint16.dpx')
```

**Comment by lgritz** [comment]
Investigating

**Comment by lgritz** [comment]
Rather than doing the `copy`, another workaround is to take advantage of the optional parameter to `write` (which makes `set_write_format` obsolete):

    buf.write('uint16.dpx', dtype='uint16')

Still, that's no excuse for a crash when using set_write_format. I can reproduce the crash. Looking into it.

**Comment by apetrynet** [comment]
Thanks! Yes, that works as well. 
I actually had that in there earlier, but strayed away from it for another reason perhaps unrelated but worth mentioning (_I hope_).

My intent with the UINT16 data type is to produce movie files with ffmpeg without color shift. I found out that the conversion in ffmpeg from float pixel type produced a slight shift while uint16 didn't. So, I first attempted to produce UINT16 exr's with `buf.write('uint16.exr', dtype='uint16')`, but it refused to produce one. I then tried `buf.set_write_format(oiio.UINT16)`, but still no UINT16 exr, so I switched to dpx and got the segfault.

**Comment by lgritz** [comment]
The OpenEXR file format does not support 16 bit (or 8 bit, for that matter) integer pixel data.

**Comment by apetrynet** [comment]
Yeah, that's what suspected. Thanks!

**Comment by lgritz** [comment]
Yeah, OpenEXR only supports 'float' (32 bit), 'half' (16 bit float), and unsigned 32 bit integer (generally only used for object ID values).

Any time you ask OIIO to write a data type that the file format itself cannot support, it instead will automatically choose a data type that is supported by the file that loses as little precision and range as possible (compared to what you requested). For example, if you ask for uint16 but OpenEXR can't accommodate it, it will instead write float values to the file (since every possible uint16 value is representable as float, but not as half).

**Comment by lgritz** [comment]
I found the bug that was leading to the crash. Will post a PR soon (or maybe tomorrow morning, it's getting late).

**Comment by lgritz** [comment]
Proposed fix in #2885 

It actually wasn't the conversion to uint16 per se. It was a series of unfortunate events that happened when ImageBuf::set_write_format was used for a file format that does not support writing differing per-channel data types.

**Comment by apetrynet** [comment]
Awesome! Thanks you so much! 
Now get some sleep!

**Comment by lgritz** [comment]
Closed by #2885
